EIPS/eip-3779.md

```diff
@@ -133,281 +89,196 @@ The remaining three states we can check at *validation time*:
 * invalid jump, and
 * invalid instruction.
 
-That is to say:
-> Valid contracts will not halt with an exception unless they either
-> * throw `out of gas` or
-> * recursively overflow stack.
-
 #### *Constraints on Valid Code*
 
-* Every instruction is valid.
-* Every jump is valid:
-   * Every`JUMP` and `JUMPI` is *static*.
-   * No `JUMP`, `JUMPI`, `RJUMP`, `RJUMPI`, or `RJUMPSUB` addresses immediate data.
-* The stacks are always valid:
-   * The _number_ of items on the `data stack` is always positive, and at most 1024.
-   * The _number_ of items on the `return stack` is always positive, and at most 1024.
-* The data stack is consistently aligned:
-   * The _number_ of items on the `data stack` between the current `stack pointer` and the `stack pointer` on entry to the most recent basic block is the same for each _execution_ of a _byte_code_.
-
-We define a `JUMP` or `JUMPI` instruction to be *static* if its `jumpsrc` argument was first placed on the stack via a `PUSH…` and that value has not changed since, though it may have been copied via a `DUP…` or `SWAP…`.
+* Every instruction MUST be valid.
+* Every jump MUST be valid:
+   * The `RJUMP`, `RJUMPI`, or `RJUMPSUB` instructions MUST NOT address immediate data.
+* The stacks MUST always be valid:
+   * The number of items on the `data stack` MUST always be positive, and at most 1024.
+   * The number of items on the `return stack`MUST always be positive, and at most 1024.
+* The data stack MUST be consistently aligned:
+   * The number of items on the `data stack` between the current `stack pointer` and the `stack pointer` on entry to the most recent subroutine MUST be the same for each _execution_ of a _byte_code_.
 
-The `RJUMP`, `RJUMPI` and `RJUMPSUB`instructions take their destination as an immediate argument, so they are *static*.
+The `RJUMP`, `RJUMPI` and `RJUMPSUB` instructions take their destination as an immediate argument, which cannot change at runtime.  Having constant destinations for all jumps means that all jump destinations can be validated at initialization time, not runtime.  And requiring a consistently aligned`data stack` prevents stack underflow. It can also catch such errors as misaligned stacks due to irreducible control flows and calls to subroutines with the wrong number of arguments.
```

Comment:
```suggestion
The `RJUMP`, `RJUMPI`, and `RJUMPSUB` instructions take their destination as an immediate argument, which cannot change at runtime. Having constant destinations for all jumps means that all jump destinations can be validated at initialization time, not runtime. And requiring a consistently aligned `data stack` prevents stack underflow. It can also catch such errors as misaligned stacks due to irreducible control flows and calls to subroutines with the wrong number of arguments.
```